crates/bevy_ecs/src/error/handler.rs

```diff
@@ -1,75 +1,154 @@
+        /// The name of the command that failed.
+        name: Cow<'static, str>,
+    },
+    /// The error occurred in an observer.
+    Observer {
+        /// The name of the observer that failed.
+        name: Cow<'static, str>,
+        /// The last tick that the observer was run.
+        last_run: Tick,
+    },
 }
 
-/// The default systems error handler stored as a resource in the [`World`](crate::world::World).
-pub struct DefaultSystemErrorHandler(pub fn(BevyError, SystemErrorContext));
+impl Display for ErrorContext {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::System { name, .. } => {
+                write!(f, "System `{}` failed", name)
+            }
+            Self::Command { name } => write!(f, "Command `{}` failed", name),
+            Self::Observer { name, .. } => {
+                write!(f, "Observer `{}` failed", name)
+            }
+        }
+    }
+}
 
-impl Resource for DefaultSystemErrorHandler {}
+impl ErrorContext {
+    /// The name of the ECS construct that failed.
+    pub fn name(&self) -> &str {
+        match self {
+            Self::System { name, .. }
+            | Self::Command { name, .. }
+            | Self::Observer { name, .. } => name,
+        }
+    }
 
-impl Default for DefaultSystemErrorHandler {
-    fn default() -> Self {
-        Self(panic)
+    /// A string representation of the kind of ECS construct that failed.
+    ///
+    /// This is a simpler helper used for logging.
+    pub fn kind(&self) -> &str {
+        match self {
+            Self::System { .. } => "system",
+            Self::Command { .. } => "command",
+            Self::Observer { .. } => "observer",
+        }
     }
 }
 
+/// A global error handler. This can be set at startup, as long as it is set before
+/// any uses. This should generally be configured _before_ initializing the app.
+///
+/// This should be set inside of your `main` function, before initializing the Bevy app.
+///
+/// # Examples
+///
+/// To set the default, fallback error handler:
+///
+/// ```
+/// # use bevy_ecs::error::{GLOBAL_ERROR_HANDLER, warn};
+/// GLOBAL_ERROR_HANDLER.set(warn).expect("The error handler can only be set once, globally.");
+/// // initialize Bevy App here
+/// ```
+///
+/// To use this error handler in your app for custom error handling logic:
+///
+/// ```rust
+/// fn handle_errors(error: BevyError, ctx: ErrorContext) {
+///    // By convention, panicking is the default behavior for errors if the user doesn't set an error handler.
+///    // Libraries that you ship should follow this convention to avoid surprising users!
+///    let error_handler = *GLOBAL_ERROR_HANDLER.get_or_init(|| panic);
+///    error_handler(error, ctx);        
+/// }
+/// ```
+///
+/// # Warning
+///
+/// As this can *never* be overwritten, library code should never set this value.
+pub static GLOBAL_ERROR_HANDLER: OnceLock<fn(BevyError, ErrorContext)> = OnceLock::new();
```

Comment:
Thoughts on hiding this from users and having a `App::set_error_handler` fn which can be called multiple times, and then which actually sets this in `App::start`? We could even make this private and expose it only through a `get_error_handler()` function.

I think this might be closer to what bevy users expect from our API, and would eliminate the problems caused by calling it conflicting handlers or trying to change it while the app is running. Or IDK maybe we want to be able to swap it out at runtime for the editor, in which case a OnceLock isn't the right primitive either.

Reply:
That won't work for standalone bevy_ecs users: they don't have access to App. We also *really* want this to be truly global, not set per SubApp.

That said, I am down to make this a bit more opaque to give us flexibility and a nicer API. Let me see about newtyping it.

Reply:
In that case I think the current is probably fine TBH. The API works, it was just a thought.

Reply:
It does work, but I don't like it! 😂

Reply:
We could always provide a wrapper as part of `App` that hides the less nice ECS implementation from most users.

Reply:
IMO there should be one way to set this: no need to complicate things.